### ISSUE TITLE
List terms which should be avoided

### ISSUE THREAD

**Issue by vstinner** [python/devguide#605]
I replaced master/slave with parent/child or server/client in Python. I'm now replacing whitelist/blacklist with allowlist/denylist in Python:

* https://bugs.python.org/issue34605
* https://bugs.python.org/issue41521

It may be nice to have a list of terms which should be avoided in the Python projects and propose better terms:

* master/slave => parent/child, server/client, main, primary/seconday, etc.
* whitelist/blacklist => allowlist/denylist, ignore, etc.

I don't need that we have to argue in details why specific terms should be avoided, but just give a general guideline like:

> The intent is to make the Python community more welcoming and more diverse by avoiding to make some people uncomfortable.

cc @brettcannon @warsaw @willingc @Mariatta @ejodlowska

**Comment by vstinner** [comment]
Some articles about terminology:

* https://tools.ietf.org/id/draft-knodel-terminology-01.html (draft RFC): Master-slave, Blacklist-whitelist and "Other considerations"
* https://www.businessinsider.com/companies-github-terms-master-slave-blacklist-whitelist-open-source-2020-6
* https://www.marketplace.org/2020/06/17/tech-companies-update-language-to-avoid-offensive-terms/
* https://www.washingtonpost.com/opinions/2020/06/12/tech-industry-has-an-ugly-master-slave-problem/
* https://lifehacker.com/ditch-these-racist-terms-from-your-tech-vocabulary-1844041452
* etc.

I will not attempt to give an exhaustive list of articles, it's a hot topic and there are more and more articles about that!

**Comment by hugovk** [comment]
> * [tools.ietf.org/id/draft-knodel-terminology-01.html](https://tools.ietf.org/id/draft-knodel-terminology-01.html) (draft RFC): Master-slave, Blacklist-whitelist and "Other considerations"

Newest draft is v3 (July 08, 2020):

* https://tools.ietf.org/id/draft-knodel-terminology-03.html

**Comment by vstinner** [comment]
"Inclusive Chromium code": https://chromium.googlesource.com/chromium/src/+/master/styleguide/inclusive_code.md

**Comment by brettcannon** [comment]
Are the guides moving to "allowlist"/"denylist"? I've seen "allow list"/"deny list" (notice the spaces).

**Comment by vstinner** [comment]
I don't think that we have to require one replacement. We can suggest multiple alternatives. For example, first I replaced blacklist with denylist in http.cookiejar documentation, but when I read again the whole documentation, it seems like blocklist is more appropriate: https://github.com/python/cpython/pull/21826

**Comment by PaulMcMillan** [comment]
I agree that it is reasonable to enumerate terms that should be avoided. As always, the peanut gallery (itself an economically loaded term) is likely to chime in and tell us we're all wrong for trying to do the right thing, but it is important that we persist.

I've been seeing allowlist/blocklist as an accepted replacement terminology pretty broadly across my professional security-oriented network. I like it because I find it unambiguous, and "block" has slightly stronger connotations than "deny" in parallel usage. For better or worse, it's also verbally similar enough to allow for seamless substitution and general comprehension among folks who haven't thought carefully about the issue.

For "master/slave" replacement I have a personal preference for "leader/follower" in cases where it semantically fits. It's clear, makes fewer assumptions about western nuclear family structure than "parent/child", and avoids the abiguity associated with "server/client" where "client" is a potential leader but just happens to be following (in contrast to "client" in the "queries the data store" sense).

[and not to go down a rabbit hole, but "client" can also be problematic given the feudal connotations]

More generally, I think it is helpful to enumerate specific turns of phrase to be avoided, since we otherwise risk falling to an onslaught of malintentioned arguers with a goal of derailing improvements via reductio ad absurdum.

**Comment by aeros** [comment]
@vstinner 
> master/slave => parent/child, server/client, main, primary/seconday, etc.
whitelist/blacklist => allowlist/denylist, ignore, etc.

+1 For the proposed replacements. It might be a decent idea to put together a table in the devguide (perhaps in the ["Documentating Python"](https://devguide.python.org/documenting/) section). My only feedback would be to use some caution with regards to banning the usage of terms that are not widely considered to be controversial, so that we don't end up spending an infinite amount of time on this.

I'm personally of the mindset that there are definitely outdated terms that we should move away from ("master/slave" and "blacklist/whitelist" being some of the more egregious offenders), but that it's impossible to avoid making *someone* out there uncomfortable; e.g. I'd have to disagree with @PaulMcMillan about avoiding usage of "parent/child" and "server/client". If you dig deep enough, there are going to be some potential connotations with any terminology. To me at least, those connotations have to be adequately clear and severe enough to justify the time spent replacing them (and of course, the replacement should be just as clear if not more clear than the original). Otherwise, the maintenance cost starts to outweigh the benefits, especially if it is determined that the replacements are offensive to someone a few years down the road.

**Comment by brettcannon** [comment]
At work we are avoiding "master", "slave", "blacklist", and "whitelist" and I have heard the same from other companies. I think that's good for now and if other teams come out as generally unacceptable we can handle them then.

**Comment by vstinner** [comment]
FYI there was a discussion about the usagef of gendered language in the PEP 3136: https://bugs.python.org/issue41743

**Comment by Mariatta** [comment]
I agree overall on a list of inclusive languages, but I'm not sure if the devguide is the right place for it. I think perhaps this could fall within the scope for the upcoming Docs WG, to provide a guideline of acceptable terms. I would hope we use those terminologies not just for writing documentation, comments, code, but also during normal communications like in mailing lists/discussion forums.

**Comment by vstinner** [comment]
> I agree overall on a list of inclusive languages, but I'm not sure if the devguide is the right place for it.

Do you have a better place for such list?

**Comment by vstinner** [comment]
The Docs WG doesn't exist yet, so devguide can a temporary home, until it is moved to somewhere else.

**Comment by brettcannon** [comment]
Are we in such a rush to have an allowlist of terms that we can't wait for the Docs WG?

**Comment by JulienPalard** [comment]
> I don't think that we have to require one replacement. We can suggest multiple alternatives.

Agree, as a french speaker I'm glad when alternatives are proposed (building it myself in my brain take efforts, I miss some, and it take time). Multiple alternatives is good too, so we can pick the most relevant one according to the context (which is often more semantically right than the avoided terms).

> The Docs WG doesn't exist yet, so devguide can a temporary home, until it is moved to somewhere else.

Agree, the https://devguide.python.org/documenting/ is a good place for those recommendations. (nobody denies the Docs WG to edit this page), it's not like we're creating a whole new thing.

**Comment by SpheeresX** [comment]
"parent/child" might be offensive for people who can't have children.

**Comment by 8vasu** [comment]
Dear maintainers, please consider [this comment that I made](https://github.com/python/cpython/issues/85984#issuecomment-1902055635) seriously before making any final naming decisions.

**Comment by vstinner** [comment]
"master" and "slave" PTY/FD are names are not part of the Python `pty` module, so you're free to change these names in the documentation without introducing any backward incompatible change, no?

**Comment by 8vasu** [comment]
I am not taking about backward incompatibility. I am just saying rename them to something unique; something that is neither too common nor used to refer to any other aspect of programming like processes, networking, etc.

For example, the `forkpty()` function involves both a pty pair and a pair of parent/child processes. If, for example, you rename master/slave to parent/child, not only will you introduce confusion while reading the source of the `pty` module and the `pty`-related functions in `posixmodule.c`, it will no longer be possible to quickly find all instances of `pty`-related code by performing a simple grep for `master`/`slave`.

**Comment by vstinner** [comment]
> you introduce confusion while reading the source of the pty module and the pty-related functions in posixmodule.c

Why not changing terms in both modules?

**Comment by 8vasu** [comment]
Of course you would change them exhaustively throughout the whole repository, but my request to you is to just be creative and come up with new, unique terms that to not collide/coincide/clash with existing terminology (with the only constraint being that they must start with 'm' and 's').

I suggested *mother* and *son*, but they do not have to be exactly that; you can do whatever you wish (you're the boss); just please invent NEW, unused terms for the pty pair nomenclature.

**Comment by vstinner** [comment]
Apparently, file descriptors returned by openpty() are not equal in permissions / behavior. For example, according to test_pty, apparently closing the first file descriptor can raise SIGHUP signal if the process is the session leader. If the first one is more important and/or "control" the second one, we can use the terms that I proposed in my first message:

> master/slave => parent/child, server/client, main, primary/seconday, etc.

Server/client terms are common in networking. primary/seconday is more common on databases when a secondary replicates the primary.

Here, parent/child sounds like good terms: *parent_fd* and *child_fd*.

> I suggested mother and son

I didn't see usage of these terms previously. *Parent* sounds more (gender) neutral than *mother*.

**Comment by 8vasu** [comment]
> Here, parent/child sounds like good terms: parent_fd and child_fd.

While they do sound more on-topic compared to *server/client* or *primary/secondary*, the inter-process communication terms [*master* and *slave*](https://en.wikipedia.org/wiki/Master%E2%80%93slave_(technology)#cite_note-25) do not imply a parent-child relationship between the processes using them.
1. Examples where the terms will mislead a reader to assume a false hierarchical relationship between processes: `screen(1)` and `tmux(1)` can attach or detach independently started processes to the *slave end* of a pty.
2. Examples where the *parent-child* terminology for the pty ends might be a good fit: [`forkpty()`](https://linux.die.net/man/3/forkpty), [`script(1)`](https://man7.org/linux/man-pages/man1/script.1.html); see image below.

Also, the terms *parent* and *child* do not start with *m*, *s*. It will be difficult to detect existing code that simply uses these initials as variable names.

> I didn't see usage of these terms previously. Parent sounds more (gender) neutral than mother.

You do not have to use my terms; will it be possible to get opinions of well-known system programmers? Or maybe start a poll if a democratic process is preferred?

If we use unique/new/non-conflicting terms in the source, it will be easier to change them in the future if projects other than Python reach a consensus on what terms to use and such terms become part of some standard. 

![script(1)](https://upload.wikimedia.org/wikipedia/commons/thumb/6/62/Termios-script-diagram.svg/710px-Termios-script-diagram.svg.png)

**Comment by vstinner** [comment]
I'm not sure of what you are suggesting. Which terms do you prefer?

**Comment by 8vasu** [comment]
I asked GPT4 API the following:

> We are looking for terms that can replace the terms 'master' and 'slave' used to refer to pseudoterminal ends. These are our constraints:
>
> 1. The terms must be inclusive.
> 2. The terms must retain the initials 'm' and 's'.
> 3. The terms must still convey the message that the master end is the "control center" of the slave end.
> 4. The terms must **not** coincide with existing technology terms or other common terms such as parent/child, server/client, first/second, main/subordinate, primary/secondary, etc.

Its outputs are at the bottom. I like `modulator` and `satellite`, which `grep(1)` finds only 3 and 0 times in the Python repository respectively.

@vstinner Note: you probably know this already, but just in case yo do not: in your root comment (and in subsequent comments that copied it), it says "seconday" instead of "seconda**R**y".

---

- `manager` and `satellite`
- `monitor` and `surrogate`
- `maestro` and `satellite`
- `mentor` and `scholar`
- `mainstay` and `supporter`
- `mediator` and `subscriber`
- `magistrate` and `steward`
- `marshal` and `squire`
- `major` and `sentry`
- `mastermind` and `synergist`
- `modulator` and `sequencer`
- `mainframe` and `satellite`
- `multiplexer` and `signal`
- `matrix` and `sensor`

**Comment by vstinner** [comment]
@8vasu: I don't think that the devguide issue tracker is the right place to discuss very specific names of the Python pty module. Sorry, I shouldn't have replied here. I suggest to continue the discussion in the Python bug tracker.

**Comment by 8vasu** [comment]
I understand.
